examples/ex22.cpp

```diff
@@ -351,48 +362,48 @@ int main(int argc, char *argv[])
    u = 0.0;
    U = 0.0;
 
-   OperatorHandle PCOp;
-   pcOp->FormSystemMatrix(ess_tdof_list, PCOp);
-
-   {
-      ComplexSparseMatrix * Asp =
-         dynamic_cast<ComplexSparseMatrix*>(A.Ptr());
-
-      cout << "Size of linear system: "
-           << 2 * Asp->real().Width() << endl << endl;
-   }
+   cout << "Size of linear system: " << A.Ptr()->Width() << endl << endl;
 
    // 10. Define and apply a GMRES solver for AU=B with a block diagonal
    //     preconditioner based on the appropriate sparse smoother.
    {
       Array<int> blockOffsets;
       blockOffsets.SetSize(3);
       blockOffsets[0] = 0;
-      blockOffsets[1] = PCOp.Ptr()->Height();
-      blockOffsets[2] = PCOp.Ptr()->Height();
+      blockOffsets[1] = A.Ptr()->Height() / 2;
+      blockOffsets[2] = A.Ptr()->Height() / 2;
       blockOffsets.PartialSum();
 
       BlockDiagonalPreconditioner BDP(blockOffsets);
 
       Operator * pc_r = NULL;
       Operator * pc_i = NULL;
 
-      double s = 1.0;
-      switch (prob)
+      if (pa)
       {
-         case 0:
-            pc_r = new DSmoother(*PCOp.As<SparseMatrix>());
-            break;
-         case 1:
-            pc_r = new GSSmoother(*PCOp.As<SparseMatrix>());
-            s = -1.0;
-            break;
-         case 2:
-            pc_r = new DSmoother(*PCOp.As<SparseMatrix>());
-            break;
-
-         default: break; // This should be unreachable
+         pc_r = new OperatorJacobiSmoother(*pcOp, ess_tdof_list);
+      }
+      else
+      {
+         OperatorHandle PCOp;
+         pcOp->SetDiagonalPolicy(mfem::Operator::DIAG_ONE);
+         pcOp->FormSystemMatrix(ess_tdof_list, PCOp);
```

Comment:
I have to admit that the memory management here is a little bit confusing (but this is not specific to this example).
`FormSystemMatrix(ess_tdof_list, PCOp)` will make `PCOp` point to a `SparseMatrix` that is internal to the bilinear form, which will persist even when `PCOp` goes out of scope.

Someone reading the code who is less familiar with the internals of the bilinear form may suspect that the sparse matrix will be destroyed when the operator handle goes out of scope, and that would be a problem because `DSmoother` and `GSSmoother` hold pointers to the sparse matrix.

I don't think this needs to be changed here, and the code looks correct to me, but I did find it a bit confusing...